invenio_resources/service/service.py

```diff
@@ -0,0 +1,189 @@
+
+    record_cls = Record
+    resolver_cls = Resolver
+    resolver_obj_type = "rec"
+    pid_type = "recid"  # PID type for resolver and minter
+    permission_policy_cls = ""
+    record_state_cls = RecordState
+    # Class dependency injection
+    indexer_cls = RecordIndexer
+    search_cls = RecordsSearch
+    search_engine_cls = SearchEngine
+
+    # pid = 'recidv2'
+    # pids = ['doi', ]
+
+    #
+    # search_class = None
+    # query_interpreter_class = None
+    # state_class = None
+    # draft_class = None
+    # result_class = None
+    # action_classes = {
+    #     'publish': ...
+    # }
+
+
+class RecordServiceFactory:
+    """Factory for creating object instances and classes."""
+
+    def __init__(self, config):
+        """Constructor."""
+        self._config = config
+
+    def resolver(self):
+        """Factory for creating a resolver class."""
+        return self._config.resolver_cls(
+            pid_type=self._config.pid_type,
+            getter=self._config.record_cls.get_record,
+        )
+
+    def minter(self):
+        """Factory for creating a minter class."""
+        return current_pidstore.minters[self._config.pid_type]
+
+    def fetcher(self):
+        """Factory for creating a fetcher class."""
+        return current_pidstore.fetchers[self._config.pid_type]
+
+    def indexer(self):
+        """Factory for creating a indexer class."""
+        return self._config.indexer_cls()
+
+    def search(self):
+        """Factory for creating a search class."""
+        return self._config.search_engine_cls(self._config.search_cls)
+
+    def permission(self, action_name, **kwargs):
+        """Factory for creating permissions from a permission policy."""
+        if self._config.permission_policy_cls:
+            return self._config.permission_policy_cls(action_name, **kwargs)
+        else:
+            return Permission()
+
+    def record(self):
+        """Factory for creating a record class."""
+        return self._config.record_cls
+
+    def record_state(self, **kwargs):
+        """Create a new item state."""
+        return self._config.record_state_cls(**kwargs)
+
+
+class RecordService:
+    """Record Service interface."""
+
+    factory = RecordServiceFactory(RecordServiceConfig)
+
+    #
+    # Permissions checking
+    #
+    @classmethod
+    def require_permission(cls, identity, action_name, **kwargs):
+        """Require a specific permission from the permission policy."""
+        if not cls.factory.permission(action_name, **kwargs).allows(identity):
+            raise PermissionDeniedError(action_name)
+
+    #
+    # Persistent identifier resolution
+    #
+    @classmethod
+    def resolve(cls, id_):
+        """Resolve a persistent identifier to a record."""
+        return cls.factory.resolver().resolve(id_)
+
+    #
+    # High-level API
+    #
+    @classmethod
+    def get(cls, id_, identity):
+        """Retrieve a record."""
+        pid, record = cls.resolve(id_)
```

Comment:
Normal: we could have the `resolve` method just return the `State` (`IdentifiedRecord` plug-plug :wink: ). `cls` has everything needed.

Reply:
Encapsulation sounds good to me, as long as is done from `resolve` method here so we dont have to touch resolver in pidstore.